File: python-sdk/src/mlxmesh/wallet.py
"""Portable, recoverable account identity — the client-side half of
`internal/wallet.Manager`'s challenge/response protocol.

This is NOT an on-chain wallet: credits live in the coordinator's ledger
(server-authoritative internal credits, no token/chain). What `Wallet`
provides is an Ed25519 keypair that PROVES ownership of a ledger balance, so
the same account works from any device/process and is recoverable — the
coordinator never generates or sees the private key, only the derived
address and, once a challenge is signed, the public key.

Wire format matches `internal/wallet/wallet.go` and `cmd/coordinator/main.go`'s
`POST /account/challenge` / `POST /account/auth` handlers exactly:

  - address       = "oim" + hex(sha256(raw 32-byte Ed25519 public key))
  - signed message = f"oim-account-auth:{address}:{nonce}" (UTF-8 bytes)
  - public_key/signature on the wire are standard (padded) base64 — NOT
    URL-safe base64.
"""

from __future__ import annotations

import base64
import hashlib
import json
import os
from pathlib import Path

from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PrivateKey

ADDRESS_PREFIX = "oim"

DEFAULT_WALLET_PATH = Path.home() / ".config" / "mlxmesh" / "wallet.json"


def address_from_public_key(public_key_raw: bytes) -> str:
    """Mirrors wallet.AddressFromPubKey exactly: AddressPrefix + hex(sha256(pubkey))."""
    return ADDRESS_PREFIX + hashlib.sha256(public_key_raw).hexdigest()


class Wallet:
    """An Ed25519 account keypair plus its derived address.

    The private key never leaves this object except to sign — it is never
    sent over the network. Only the address and (during authentication) the
    raw public key and a signature are ever transmitted.
    """

    def __init__(self, seed: bytes):
        if len(seed) != 32:
            raise ValueError(f"wallet seed must be 32 bytes, got {len(seed)}")
        self._private_key = Ed25519PrivateKey.from_private_bytes(seed)
        self._seed = seed
        self._public_key_raw = self._private_key.public_key().public_bytes_raw()
        self.address = address_from_public_key(self._public_key_raw)

    @classmethod
    def create(cls) -> "Wallet":
        """Generates a brand-new random keypair. Nothing is persisted until
        you call .save() — the caller decides whether/where to keep it."""
        seed = os.urandom(32)
        return cls(seed)

    @classmethod
    def load(cls, path: str | Path | None = None) -> "Wallet":
        """Loads a previously-saved wallet. Raises FileNotFoundError if the
        path doesn't exist — use load_or_create() if you want silent
        first-run creation instead."""
        p = Path(path) if path is not None else DEFAULT_WALLET_PATH
        with open(p, "r", encoding="utf-8") as f:
            data = json.load(f)
        seed = bytes.fromhex(data["seed"])
        return cls(seed)

    @classmethod
    def load_or_create(cls, path: str | Path | None = None) -> "Wallet":
        """The common case: use the wallet on disk if one exists, otherwise
        generate a fresh one and persist it immediately so the address is
        stable across process restarts from the very first run."""
        p = Path(path) if path is not None else DEFAULT_WALLET_PATH
        if p.exists():
            return cls.load(p)
        wallet = cls.create()
        wallet.save(p)
        return wallet

    def save(self, path: str | Path | None = None) -> None:
        """Persists the seed as {"seed": "<64 hex chars>"} with 0600
        permissions — mirrors nodeconfig.Save's convention on the Go side,
        scoped to this SDK's own config namespace."""
        p = Path(path) if path is not None else DEFAULT_WALLET_PATH
        p.parent.mkdir(parents=True, exist_ok=True)
        with open(p, "w", encoding="utf-8") as f:
            json.dump({"seed": self._seed.hex()}, f)
        os.chmod(p, 0o600)

    def sign(self, message: bytes) -> bytes:
        """Raw Ed25519 signature over message (64 bytes)."""
        return self._private_key.sign(message)

    @property
    def public_key_b64(self) -> str:
        """Standard (padded) base64 of the raw 32-byte public key — the exact
        encoding POST /account/auth expects in its public_key field."""
        return base64.b64encode(self._public_key_raw).decode("ascii")

    def __repr__(self) -> str:
        return f"Wallet(address={self.address!r})"

File: swift-sdk/Sources/MeshKit/Wallet.swift
import CryptoKit
import Foundation

/// Portable, recoverable account identity — the client-side half of
/// `internal/wallet.Manager`'s challenge/response protocol. Mirrors
/// `OIMDashboard/iOS/Wallet/WalletStore.swift`'s keypair/address logic
/// exactly, but with pluggable storage rather than one app's hardcoded
/// Keychain access group — MeshKit is a library other apps embed.
///
/// Wire format matches `internal/wallet/wallet.go` and
/// `cmd/coordinator/main.go`'s `POST /account/challenge` / `POST
/// /account/auth` handlers exactly:
///   - address        = "oim" + hex(sha256(raw 32-byte Ed25519 public key))
///   - signed message  = "oim-account-auth:{address}:{nonce}" (UTF-8 bytes)
///   - public_key/signature on the wire are standard (padded) base64.
public struct Wallet: Sendable {
    public let address: String
    /// The raw 32-byte Ed25519 seed — stored as plain Data (not a live
    /// CryptoKit key object) so Wallet stays trivially Sendable for use from
    /// the MeshClient actor. Never sent over the network; only used locally
    /// to sign and to derive the public key.
    private let seed: Data

    public init(seed: Data) throws {
        guard seed.count == 32 else {
            throw WalletError.invalidSeedLength(seed.count)
        }
        self.seed = seed
        let key = try Curve25519.Signing.PrivateKey(rawRepresentation: seed)
        self.address = Wallet.deriveAddress(publicKey: key.publicKey)
    }

    /// Generates a brand-new random keypair. Nothing is persisted until you
    /// call .save() — the caller decides whether/where to keep it.
    public static func create() -> Wallet {
        let key = Curve25519.Signing.PrivateKey()
        // A freshly-generated Curve25519 private key's rawRepresentation is
        // always exactly 32 bytes — this cannot actually fail.
        return try! Wallet(seed: key.rawRepresentation)
    }

    private static func deriveAddress(publicKey: Curve25519.Signing.PublicKey) -> String {
        let digest = SHA256.hash(data: publicKey.rawRepresentation)
        return "oim" + digest.map { String(format: "%02x", $0) }.joined()
    }

    /// Raw Ed25519 signature over message (64 bytes).
    public func sign(_ message: Data) throws -> Data {
        let key = try Curve25519.Signing.PrivateKey(rawRepresentation: seed)
        return try key.signature(for: message)
    }

    /// Standard (padded) base64 of the raw 32-byte public key — the exact
    /// encoding POST /account/auth expects in its public_key field.
    public var publicKeyBase64: String {
        let key = try! Curve25519.Signing.PrivateKey(rawRepresentation: seed)
        return key.publicKey.rawRepresentation.base64EncodedString()
    }

    // MARK: - Persistence

    public func save(storage: WalletStorage = FileWalletStorage()) throws {
        try storage.save(seed: seed)
    }

    /// Loads a previously-saved wallet. Throws WalletError.notFound if there
    /// isn't one — use loadOrCreate() for silent first-run creation instead.
    public static func load(storage: WalletStorage = FileWalletStorage()) throws -> Wallet {
        guard let seed = try storage.load() else {
            throw WalletError.notFound
        }
        return try Wallet(seed: seed)
    }

    /// The common case: use the wallet in storage if one exists, otherwise
    /// generate a fresh one and persist it immediately so the address is
    /// stable across process restarts from the very first run.
    public static func loadOrCreate(storage: WalletStorage = FileWalletStorage()) throws -> Wallet {
        if let seed = try storage.load() {
            return try Wallet(seed: seed)
        }
        let wallet = Wallet.create()
        try wallet.save(storage: storage)
        return wallet
    }
}

public enum WalletError: Error, LocalizedError {
    case invalidSeedLength(Int)
    case notFound

    public var errorDescription: String? {
        switch self {
        case .invalidSeedLength(let n): return "wallet seed must be 32 bytes, got \(n)"
        case .notFound: return "no wallet found at the configured storage location"
        }
    }
}

/// Pluggable persistence for a wallet's seed. MeshKit ships a file-based
/// default (FileWalletStorage) that works on all four supported platforms
/// without assuming Keychain or any one app's access group — embedding apps
/// that want Keychain-backed storage (as OIMDashboard's own WalletStore
/// does) supply their own conformance.
public protocol WalletStorage: Sendable {
    func save(seed: Data) throws
    func load() throws -> Data?
}

/// Default storage: a JSON file `{"seed": "<64 hex chars>"}` under the
/// platform's Application Support directory, POSIX permissions 0600.
public struct FileWalletStorage: WalletStorage {
    private let path: URL

    public init(path: URL? = nil) {
        if let path {
            self.path = path
        } else {
            let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first
                ?? FileManager.default.temporaryDirectory
            self.path = base.appendingPathComponent("mlxmesh", isDirectory: true).appendingPathComponent("wallet.json")
        }
    }

    public func save(seed: Data) throws {
        let dir = path.deletingLastPathComponent()
        try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        let hex = seed.map { String(format: "%02x", $0) }.joined()
        let json = try JSONSerialization.data(withJSONObject: ["seed": hex])
        try json.write(to: path, options: .atomic)
        try FileManager.default.setAttributes([.posixPermissions: 0o600], ofItemAtPath: path.path)
    }

    public func load() throws -> Data? {
        guard FileManager.default.fileExists(atPath: path.path) else { return nil }
        let data = try Data(contentsOf: path)
        guard let obj = try JSONSerialization.jsonObject(with: data) as? [String: String],
              let hex = obj["seed"], let seed = Data(hexEncoded: hex)
        else {
            throw WalletError.notFound
        }
        return seed
    }
}

extension Data {
    fileprivate init?(hexEncoded hex: String) {
        guard hex.count % 2 == 0 else { return nil }
        var data = Data(capacity: hex.count / 2)
        var idx = hex.startIndex
        while idx < hex.endIndex {
            let next = hex.index(idx, offsetBy: 2)
            guard let byte = UInt8(hex[idx..<next], radix: 16) else { return nil }
            data.append(byte)
            idx = next
        }
        self = data
    }
}
